File: zig/bubble_sort.zig
const std = @import("std");
const print = std.debug.print;

fn bubbleSort(arr: []i32) void {
    const n = arr.len;
    var i: usize = 0;
    while (i < n - 1) : (i += 1) {
        var j: usize = 0;
        while (j < n - i - 1) : (j += 1) {
            if (arr[j] > arr[j + 1]) {
                const temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
            }
        }
    }
}

fn generateArray(allocator: std.mem.Allocator, n: usize) ![]i32 {
    const arr = try allocator.alloc(i32, n);
    var seed: u32 = 42; // Fixed seed for reproducible results
    
    for (arr) |*item| {
        seed = seed *% 1103515245 +% 12345;
        item.* = @intCast(seed % 10000);
    }
    return arr;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();
    
    const args = try std.process.argsAlloc(allocator);
    defer std.process.argsFree(allocator, args);
    
    if (args.len < 2) {
        print("Usage: bubble_sort <n>\n", .{});
        return;
    }
    
    const n = try std.fmt.parseInt(usize, args[1], 10);
    const arr = try generateArray(allocator, n);
    defer allocator.free(arr);
    
    const start = std.time.nanoTimestamp();
    bubbleSort(arr);
    const end = std.time.nanoTimestamp();
    
    const time_ms = @divFloor(end - start, 1_000_000);
    
    print("Zig: bubble_sort({d}) = sorted\n", .{n});
    print("Time: {d}ms\n", .{time_ms});
}

File: java/bubble_sort.java
import java.util.Random;

public class bubble_sort {
    
    public static void bubbleSort(int[] arr) {
        int n = arr.length;
        for (int i = 0; i < n - 1; i++) {
            for (int j = i + 1; j < n; j++) {
                if (arr[i] > arr[j]) {
                    int temp = arr[j];
                    arr[j] = arr[i];
                    arr[i] = temp;
                }
            }
        }
    }
    
    public static int[] generateArray(int n) {
        Random random = new Random(42); // Fixed seed for reproducible results
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = random.nextInt(10000);
        }
        return arr;
    }
    
    public static void main(String[] args) {
        if (args.length < 1) {
            System.out.println("Usage: BubbleSort <n>");
            return;
        }
        
        int n = Integer.parseInt(args[0]);
        int[] arr = generateArray(n);
        
        long start = System.currentTimeMillis();
        bubbleSort(arr);
        long end = System.currentTimeMillis();
        
        long timeMs = end - start;
        
        System.out.println("Java: bubble_sort(" + n + ") = sorted");
        System.out.println("Time: " + timeMs + "ms");
    }
}